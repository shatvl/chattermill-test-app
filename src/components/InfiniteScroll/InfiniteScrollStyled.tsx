import styled, { keyframes } from 'styled-components';

export const loadingAnimation = keyframes`
  0% {
    font-size: 1.5rem;
  }
  50% {
    font-size: 2.5rem;
  }
  100% {
    font-size: 1.5rem;
  }
`;

export const InfiniteScrollWrapper = styled.div`
  overflow-y: auto;
`;

export const InfiniteScrollAnchor = styled.div``;

export const InfiniteScrollLoader = styled.div`
  width: 100%;
  height: 2.5rem;
  margin-bottom: 0.75rem;
  text-align: center;
  position: relative;
  animation: ${loadingAnimation} 1.2s linear infinite;
`;
