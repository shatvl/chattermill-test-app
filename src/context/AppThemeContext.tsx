import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
