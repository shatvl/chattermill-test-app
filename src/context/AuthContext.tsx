import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
