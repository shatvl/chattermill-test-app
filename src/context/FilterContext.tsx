import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
