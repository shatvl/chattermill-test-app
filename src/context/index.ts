export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
