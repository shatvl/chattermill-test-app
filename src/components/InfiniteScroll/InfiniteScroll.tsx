import React, { FC, useEffect, useRef, useState } from 'react';
import { InfiniteScrollAnchor, InfiniteScrollLoader, InfiniteScrollWrapper } from './InfiniteScrollStyled';

interface InifiniteScrollProps {
  hasMore: boolean;
  loading: boolean;
  load: () => void;
}

const InfiniteScroll: FC<InifiniteScrollProps> = ({ hasMore, load, loading, children }) => {
  // acnchor to observe appearence in viewport
  const anchorRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const observer = new IntersectionObserver(
      ([entry]: IntersectionObserverEntry[]) => {
        if (entry.isIntersecting && !loading && hasMore) {
          load();
        }
      },
      {
        root: null,
        rootMargin: '0px',
        threshold: 1,
      },
    );

    if (anchorRef?.current) {
      observer.observe(anchorRef.current);
    }

    // disconnect from all .observes
    return () => {
      observer.disconnect();
    };
  }, [anchorRef]);

  return (
    <InfiniteScrollWrapper>
      {children}
      <InfiniteScrollAnchor ref={anchorRef} />
      {loading && <InfiniteScrollLoader>ðŸ˜´</InfiniteScrollLoader>}
    </InfiniteScrollWrapper>
  );
};

export default InfiniteScroll;
