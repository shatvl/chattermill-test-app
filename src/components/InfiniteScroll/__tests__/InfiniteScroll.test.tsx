/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { mount } from 'enzyme';
import { wait, waitRenderer } from '../../../testing/wait';
import { AppThemeProvider } from '../../../context/AppThemeContext';
import InfiniteScroll from '../InfiniteScroll';
import { InfiniteScrollLoader } from '../InfiniteScrollStyled';

describe('renders infinite scroll', () => {
  const mockIntersectionObserver = jest.fn();
  const mockObserve = jest.fn();
  const mockDisconnect = jest.fn();

  beforeEach(() => {
    mockIntersectionObserver.mockReturnValue({
      observe: mockObserve,
      unobserve: () => null,
      disconnect: mockDisconnect,
    });
    window.IntersectionObserver = mockIntersectionObserver;
  });

  const loadFn = jest.fn();

  it('infinite scroll snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <InfiniteScroll loading={false} hasMore={true} load={loadFn}></InfiniteScroll>
      </AppThemeProvider>,
    );

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('infinite scroll with loading snapshot', async () => {
    const component = mount(
      <AppThemeProvider>
        <InfiniteScroll loading={true} hasMore={true} load={loadFn}></InfiniteScroll>
      </AppThemeProvider>,
    );

    await wait();

    component.update();

    expect(component.find(InfiniteScrollLoader)).toBeDefined();
  });

  it('infinite scroll IntersectionObserver API are called', async () => {
    const component = mount(
      <AppThemeProvider>
        <InfiniteScroll loading={true} hasMore={true} load={loadFn}></InfiniteScroll>
      </AppThemeProvider>,
    );

    await wait();

    component.update();

    expect(mockObserve).toBeCalled();
    expect(mockDisconnect).toBeCalledTimes(0);

    component.unmount();
    expect(mockDisconnect).toBeCalledTimes(1);
  });
});
