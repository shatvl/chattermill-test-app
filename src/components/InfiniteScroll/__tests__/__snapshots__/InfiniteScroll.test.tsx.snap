// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders infinite scroll infinite scroll snapshot 1`] = `
.c0 {
  overflow-y: auto;
}

<div
  className="c0"
>
  <div
    className=""
  />
</div>
`;
